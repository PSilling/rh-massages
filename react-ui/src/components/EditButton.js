// react imports
import React, { Component } from 'react';

/**
 * Custom edit button component.
 */
class EditButton extends Component {

  render() {
    return(
      <span>
        <button style={{ 'color': '#000' }}  type="button" className="btn btn-link" onClick={this.props.onEdit}>
          <span className="glyphicon glyphicon-pencil"></span>
        </button>
      </span>
    );
  }
}

export default EditButton
