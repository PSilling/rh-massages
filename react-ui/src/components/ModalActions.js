// react imports
import React, { Component } from 'react';

// util imports
import _t from '../utils/Translations.js';

/**
 * Custom component rendering action buttons for modal footers.
 */
class ModalActions extends Component {

  render() {
    return(
      <div className="pull-right">
        <button type="button" className="btn btn-primary" onClick={this.props.onProceed}
          style={{ 'marginRight': '5px' }} autoFocus={this.props.autoFocus}>
          {this.props.primaryLabel}
        </button>
        <button type="button" className="btn btn-default"
          onClick={this.props.onClose}>
          { _t.translate('Dismiss') }
        </button>
      </div>
    );
  }
}

export default ModalActions
