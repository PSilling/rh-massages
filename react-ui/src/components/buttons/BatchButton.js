// react imports
import React, { Component } from 'react';
import PropTypes from 'prop-types';

/**
 * A default styled button component with a given label. Used primarily for
 * batch operation buttons but can be used for any label-only button.
 */
class BatchButton extends Component {

  render() {
    return (
      <span>
        <button type="button" className={"btn btn-default" + (this.props.active ? " active" : "")}
          onClick={this.props.onClick} disabled={this.props.disabled}>
          {this.props.label}
        </button>
      </span>
    )
  }
}

BatchButton.propTypes = {
  /** function to be called on button click */
  onClick: PropTypes.func.isRequired,
   /** button label */
  label: PropTypes.string,
  /** whether the button should be disabled */
  disabled: PropTypes.bool,
  /** whether the button should be displayed as active */
  active: PropTypes.bool
};

BatchButton.defaultProps = {
  disabled: false,
  active: false
};

export default BatchButton
