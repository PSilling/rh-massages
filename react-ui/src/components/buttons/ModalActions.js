// react imports
import React from "react";
import PropTypes from "prop-types";

// util imports
import _t from "../../util/Translations";

/**
 * Footer buttons for Modal dialogs. Children components can be given to add more
 * case specific elements.
 */
const ModalActions = function ModalActions(props) {
  return (
    <div className="pull-right">
      {props.children}
      {props.primaryLabel === "none" ? (
        ""
      ) : (
        <button
          type="button"
          className="btn btn-primary"
          onClick={props.onProceed}
          title={props.title}
          style={{ marginRight: "5px" }}
          autoFocus={props.autoFocus}
          disabled={props.disabled}
        >
          {props.primaryLabel}
        </button>
      )}
      <button type="button" className="btn btn-default" onClick={props.onClose}>
        {_t.translate("Dismiss")}
      </button>
    </div>
  );
};

ModalActions.propTypes = {
  /** callback function triggered on close button click */
  onClose: PropTypes.func.isRequired,
  /** callback function triggered on primary button click */
  onProceed: PropTypes.func.isRequired,
  /** whether the primary button should be automatically focused */
  autoFocus: PropTypes.bool,
  /** extra elements to be provided before the two buttons */
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]),
  /** whether the primary button should be disabled */
  disabled: PropTypes.bool,
  /** primary button label to be displayed */
  primaryLabel: PropTypes.string,
  /** primary button title */
  title: PropTypes.string
};

ModalActions.defaultProps = {
  autoFocus: false,
  children: null,
  disabled: false,
  primaryLabel: _t.translate("Proceed"),
  title: ""
};

export default ModalActions;
