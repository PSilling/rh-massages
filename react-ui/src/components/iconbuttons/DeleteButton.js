// react imports
import React, { Component } from 'react';

// component imports
import ConfirmationModal from '../modals/ConfirmationModal';

// util imports
import _t from '../../util/Translations';

/**
 * Custom delete button component.
 */
class DeleteButton extends Component {

  state = {active: false}

  handleToggle = () => {
    this.setState({active: !this.state.active});
  }

  render() {
    return(
      <span>
        <button style={{ 'color': '#000' }} type="button" className="btn btn-link"
          onClick={this.handleToggle} title={ _t.translate("Delete") }>
          <span className="glyphicon glyphicon-remove"></span>
        </button>
        {this.state.active ?
          <ConfirmationModal
            message={ _t.translate('Are you sure? This action cannot be reverted.') }
            onClose={() => this.handleToggle()}
            onConfirm={() => {
              this.handleToggle();
              this.props.onDelete();
            }}
          /> : ''
        }
      </span>
    )
  }
}

export default DeleteButton
