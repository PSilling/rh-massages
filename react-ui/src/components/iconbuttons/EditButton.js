// react imports
import React from "react";
import PropTypes from "prop-types";

// util imports
import _t from "../../util/Translations";

/**
 * Icon only button used for element editing.
 */
const EditButton = function EditButton(props) {
  return (
    <span>
      <button
        style={{ color: "#000" }}
        type="button"
        className="btn btn-link"
        onClick={props.onEdit}
        title={_t.translate("Edit")}
      >
        <span className="glyphicon glyphicon-pencil" />
      </button>
    </span>
  );
};

EditButton.propTypes = {
  /** function to be called on button click */
  onEdit: PropTypes.func.isRequired
};

export default EditButton;
