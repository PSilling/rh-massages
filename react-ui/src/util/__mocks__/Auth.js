const isAdmin = jest.fn().mockReturnValue(true);
const isAuthenticated = jest.fn().mockReturnValue(true);
const subscribed = true;
const getToken = jest.fn().mockReturnValue("secret");
const getSub = jest.fn().mockReturnValue("test");
const getClient = jest.fn().mockReturnValue({
  sub: "test", email: "test@email.com",
  name: "Test", surname: "User",
  subscribed: true
});

module.exports = {
  isAdmin,
  isAuthenticated,
  subscribed,
  getToken,
  getSub,
  getClient
}
