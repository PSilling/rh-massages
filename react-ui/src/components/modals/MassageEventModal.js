// react imports
import React, { Component } from "react";
import PropTypes from "prop-types";

// module imports
import { Row, Col, Button, Modal, ModalBody, Tooltip } from "reactstrap";
import moment from "moment";

// component imports
import ConfirmationIconButton from "../iconbuttons/ConfirmationIconButton";
import ModalActions from "../buttons/ModalActions";
import TooltipIconButton from "../iconbuttons/TooltipIconButton";

// util imports
import Auth from "../../util/Auth";
import _t from "../../util/Translations";
import Util from "../../util/Util";

/**
 * Modal displaying details about a given Massage event.
 * Also supports assignment management.
 */
class MassageEventModal extends Component {
  state = { tooltipActive: false };

  tooltipTarget = Util.getTooltipTargets(1)[0];

  definitionClass = "mb-3";

  onCalendarConfirm = () => {
    window.open(Util.getEventLink(this.props.event.massage), "_blank");
    this.props.onConfirm();
  };

  toggleTooltip = () => {
    this.setState(prevState => ({ tooltipActive: !prevState.tooltipActive }));
  };

  handleKeyPress = event => {
    if (event.key === "Enter" && document.activeElement.tabIndex === -1) {
      this.props.onClose();
    }
  };

  createChildrenButton = () => {
    let onConfirm;
    let label;
    let tooltip;

    if (Auth.isMasseur() && this.props.event.massage.client === null) {
      return null;
    }

    if (this.props.label === _t.translate("Assign me")) {
      onConfirm = this.onCalendarConfirm;
      label = _t.translate("Assign and add to calendar");
      tooltip = this.props.disabled
        ? _t.translate("Maximal simultaneous massage time per user would be exceeded")
        : _t.translate("Assigns this massage and opens a predefined Google event editor in a new tab");
    } else if (this.props.label === _t.translate("Unassign me")) {
      onConfirm = () => {
        window.open(Util.getEventLink(this.props.event.massage), "_blank");
        this.props.onClose();
      };
      label = _t.translate("Add to Google Calendar");
      tooltip = _t.translate("Opens a predefined Google event editor in a new tab");
    }

    if (label === undefined) {
      return null;
    }
    return (
      <span title={this.props.disabled ? _t.translate("Over the limit") : ""}>
        <Button
          id={this.tooltipTarget}
          className="mr-2"
          tag="a"
          color="primary"
          disabled={this.props.disabled}
          onClick={onConfirm}
          target="_blank"
          rel="noopener noreferrer"
          tabIndex="-1"
        >
          {label}
        </Button>
        <Tooltip
          isOpen={this.state.tooltipActive}
          target={this.tooltipTarget}
          toggle={this.toggleTooltip}
          trigger="hover"
        >
          {tooltip}
        </Tooltip>
      </span>
    );
  };

  createInsides = () => (
    <ModalBody>
      <Row>
        <Col md="12">
          <h3>
            {_t.translate("Details")}
            {(Auth.isAdmin() || (Auth.isMasseur() && Auth.getSub() === this.props.event.massage.masseuse.sub)) && (
              <div className="float-right">
                {this.props.allowEditation && (
                  <TooltipIconButton icon="edit" onClick={this.props.onEdit} tooltip={_t.translate("Edit")} />
                )}
                {this.props.allowDeletion && (
                  <ConfirmationIconButton
                    className="ml-2"
                    icon="trash"
                    onConfirm={this.props.onDelete}
                    tooltip={_t.translate("Delete")}
                  />
                )}
              </div>
            )}
          </h3>
          <hr />
        </Col>
      </Row>

      <Row>
        <Col md="4">
          <dl>
            <dt>{_t.translate("Facility")}</dt>
            <dd className={this.definitionClass}>{this.props.event.massage.facility.name}</dd>
            <dt>{_t.translate("Time")}</dt>
            <dd className={this.definitionClass}>
              {`${moment(this.props.event.massage.date).format("LT")} â€“ ${moment(
                this.props.event.massage.ending
              ).format("LT")}`}
            </dd>
            <dt>{_t.translate("Client")}</dt>
            {Util.isEmpty(this.props.event.massage.client) ? (
              <dd className={this.definitionClass}>
                <p style={{ color: Util.SUCCESS_COLOR }}>
                  <strong>{_t.translate("Free")}</strong>
                </p>
              </dd>
            ) : (
              <dd className={this.definitionClass}>
                <p
                  style={{
                    color:
                      !this.props.allowDeletion ||
                      (this.props.allowEditation && this.props.event.massage.client.sub === Auth.getSub())
                        ? Util.WARNING_COLOR
                        : Util.ERROR_COLOR
                  }}
                >
                  <strong>{this.props.event.massage.client.email}</strong>
                </p>
              </dd>
            )}
          </dl>
        </Col>

        <Col md="4">
          <dl>
            <dt>{_t.translate("Masseur/Masseuse")}</dt>
            <dd className={this.definitionClass}>
              {`${this.props.event.massage.masseuse.name} ${this.props.event.massage.masseuse.surname}`}
            </dd>
            <dt>{_t.translate("Duration")}</dt>
            <dd className={this.definitionClass}>
              {`${moment
                .duration(moment(this.props.event.massage.ending).diff(this.props.event.massage.date))
                .asMinutes()} ${_t.translate("minutes")}`}
            </dd>
          </dl>
        </Col>

        <Col md="4">
          <dl>
            <dt>{_t.translate("Contact")}</dt>
            <dd className={this.definitionClass}>{this.props.event.massage.masseuse.email}</dd>
            <dt>{_t.translate("Date")}</dt>
            <dd className={this.definitionClass}>{moment(this.props.event.massage.date).format("L")}</dd>
          </dl>
        </Col>
      </Row>

      <ModalActions
        primaryLabel={Auth.isMasseur() && this.props.event.massage.client === null ? "none" : this.props.label}
        title={this.props.disabled ? _t.translate("Over the limit") : ""}
        disabled={this.props.disabled}
        onProceed={this.props.onConfirm}
        onClose={this.props.onClose}
      >
        {this.createChildrenButton()}
      </ModalActions>
    </ModalBody>
  );

  render() {
    return this.props.withPortal ? (
      <Modal size="lg" isOpen toggle={this.props.onClose} tabIndex="-1" onKeyPress={this.handleModalKeyPress}>
        {this.createInsides()}
      </Modal>
    ) : (
      this.createInsides()
    );
  }
}

MassageEventModal.propTypes = {
  /** event to be displayed */
  event: PropTypes.shape({
    bgColor: PropTypes.string,
    massage: PropTypes.shape({
      id: PropTypes.number,
      masseuse: PropTypes.shape({
        email: PropTypes.string,
        masseur: PropTypes.bool,
        name: PropTypes.string,
        sub: PropTypes.string,
        subscribed: PropTypes.bool,
        surname: PropTypes.string
      }),
      date: PropTypes.oneOfType([PropTypes.number, PropTypes.instanceOf(Date)]),
      ending: PropTypes.oneOfType([PropTypes.number, PropTypes.instanceOf(Date)]),
      client: PropTypes.shape({
        email: PropTypes.string,
        masseur: PropTypes.bool,
        name: PropTypes.string,
        surname: PropTypes.string,
        sub: PropTypes.string,
        subscribed: PropTypes.bool
      }),
      facility: PropTypes.shape({
        id: PropTypes.number,
        name: PropTypes.string
      })
    })
  }).isRequired,
  /** primary button label */
  label: PropTypes.string.isRequired,
  /** function called on modal dismissal */
  onClose: PropTypes.func.isRequired,
  /** function called on primary label action */
  onConfirm: PropTypes.func.isRequired,
  /** whether the delete button should be shown */
  allowDeletion: PropTypes.bool,
  /** whether the edit button should be shown */
  allowEditation: PropTypes.bool,
  /** whether the primary button should be disabled */
  disabled: PropTypes.bool,
  /** function called on edit button action */
  onEdit: PropTypes.func,
  /** function called on delete button action */
  onDelete: PropTypes.func,
  /** whether ModalContainer should be used; useful for testing to avoid portals */
  withPortal: PropTypes.bool
};

MassageEventModal.defaultProps = {
  allowDeletion: true,
  allowEditation: true,
  disabled: false,
  onEdit() {},
  onDelete() {},
  withPortal: true
};

export default MassageEventModal;
