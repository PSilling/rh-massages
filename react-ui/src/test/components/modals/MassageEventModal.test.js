// react imports
import React from 'react';
import TestRenderer from 'react-test-renderer';

// test imports
import Auth from '../../../util/Auth';
import BatchButton from '../../../components/buttons/BatchButton';
import DeleteButton from '../../../components/iconbuttons/DeleteButton';
import EditButton from '../../../components/iconbuttons/EditButton';
import MassageEventModal from '../../../components/modals/MassageEventModal';
import ModalActions from '../../../components/buttons/ModalActions';
import _t from '../../../util/Translations';

// test mocks
jest.mock('../../../util/Auth');

afterEach(() => {
  jest.resetAllMocks();
});

test('renders inside content with correct props', () => {
  const testConfirmFunction = jest.fn(),
        testCloseFunction = jest.fn(),
        testEditFunction = jest.fn(),
        testDeleteFunction = jest.fn(),
        testEvent = {
          bgColor: "white",
          massage : {
            id: 1,
            date: new Date(0),
            ending: new Date(1000),
            client: null,
            facility: { id: 1, name: "test" }
          }
        },
        testRenderer = TestRenderer.create(
          <MassageEventModal
            event={testEvent}
            label={ _t.translate('Assign me') }
            onConfirm={testConfirmFunction}
            onClose={testCloseFunction}
            onEdit={testEditFunction}
            onDelete={testDeleteFunction}
            withPortal={false}
          />
        ),
        testInstance = testRenderer.root;

  let editButton = testInstance.findByType(EditButton),
      deleteButton = testInstance.findByType(DeleteButton),
      actions = testInstance.findByType(ModalActions),
      definitions = testInstance.findAllByType('dd'),
      successParagraph = testInstance.findByProps({ className: "text-success" }),
      treeJSON = testRenderer.toJSON();

  expect(editButton.props.onEdit).toBe(testEditFunction);
  expect(deleteButton.props.onDelete).toBe(testDeleteFunction);
  expect(actions.props.children).not.toEqual('');
  expect(actions.props.title).toBe('');
  expect(actions.props.disabled).toBe(false);
  expect(actions.props.onProceed).toBe(testConfirmFunction);
  expect(actions.props.onClose).toBe(testCloseFunction);
  expect(definitions[0].props.children).toEqual(testEvent.massage.facility.name);
  expect(treeJSON).toMatchSnapshot();
});

test('applies correct non-editation mode changes', () => {
  const testConfirmFunction = jest.fn(),
        testCloseFunction = jest.fn(),
        testEvent = {
          bgColor: "white",
          massage : {
            id: 1,
            date: new Date(0),
            ending: new Date(1000),
            client: "test",
            facility: { id: 1, name: "test" }
          }
        },
        testRenderer = TestRenderer.create(
          <MassageEventModal event={testEvent} label="test" onConfirm={testConfirmFunction}
            onClose={testCloseFunction} withPortal={false} />
        ),
        testInstance = testRenderer.root;

    let editButtons = testInstance.findAllByType(EditButton),
        dangerParagraph = testInstance.findByProps({ className: "text-danger" });

    expect(editButtons.length).toBe(0);
});

test('hides non-admin buttons properly', () => {
  Auth.isAuthenticated = jest.fn().mockImplementation(() => true);
  Auth.isAdmin = jest.fn().mockImplementation(() => false);

  const testConfirmFunction = jest.fn(),
        testCloseFunction = jest.fn(),
        testEvent = {
          bgColor: "white",
          massage : {
            id: 1,
            date: new Date(0),
            ending: new Date(1000),
            client: null,
            facility: { id: 1, name: "test" }
          }
        },
        testRenderer = TestRenderer.create(
          <MassageEventModal event={testEvent} label="test" onConfirm={testConfirmFunction}
            onClose={testCloseFunction} withPortal={false} />
        ),
        testInstance = testRenderer.root;

  let editButtons = testInstance.findAllByType(EditButton),
      deleteButtons = testInstance.findAllByType(DeleteButton);

  expect(editButtons.length).toBe(0);
  expect(deleteButtons.length).toBe(0);
});
