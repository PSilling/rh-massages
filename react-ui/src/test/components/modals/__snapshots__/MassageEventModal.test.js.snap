// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders inside content with correct props 1`] = `
<div>
  <h2>
    Podrobnosti
    <div
      className="pull-right"
    >
      <span>
        <button
          className="btn btn-link"
          onClick={[Function]}
          style={
            Object {
              "color": "#000",
            }
          }
          title="Upravit"
          type="button"
        >
          <span
            className="glyphicon glyphicon-pencil"
          />
        </button>
      </span>
      <span>
        <button
          className="btn btn-link"
          onClick={[Function]}
          style={
            Object {
              "color": "#000",
            }
          }
          title="Smazat"
          type="button"
        >
          <span
            className="glyphicon glyphicon-trash"
          />
        </button>
        
      </span>
    </div>
  </h2>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-md-4"
    >
      <dl>
        <dt>
          Budova
        </dt>
        <dd
          style={
            Object {
              "marginBottom": "0.75em",
            }
          }
        >
          test
        </dd>
        <dt>
          Čas
        </dt>
        <dd
          style={
            Object {
              "marginBottom": "0.75em",
            }
          }
        >
          01:00–01:00
        </dd>
        <dt>
          Klient
        </dt>
        <dd
          style={
            Object {
              "marginBottom": "0.75em",
            }
          }
        >
          <p
            className="text-success"
          >
            <strong>
              Volno
            </strong>
          </p>
        </dd>
      </dl>
    </div>
    <div
      className="col-md-4"
    >
      <dl>
        <dt>
          Masér/Masérka
        </dt>
        <dd
          style={
            Object {
              "marginBottom": "0.75em",
            }
          }
        />
        <dt>
          Délka trvání
        </dt>
        <dd
          style={
            Object {
              "marginBottom": "0.75em",
            }
          }
        >
          0.016666666666666666 minut
        </dd>
      </dl>
    </div>
    <div
      className="col-md-4"
    >
      <dl>
        <dt>
          Datum
        </dt>
        <dd
          style={
            Object {
              "marginBottom": "0.75em",
            }
          }
        >
          01/01/1970
        </dd>
      </dl>
    </div>
  </div>
  <div
    className="pull-right"
  >
    <a
      href="https://www.google.com/calendar/render?action=TEMPLATE&text=Masáž&dates=19700101T000000Z/19700101T000001Z&location=Red%20Hat%20Czech%20test&details=Masér/Masérka: undefined"
      tabIndex="-1"
      target="_blank"
    >
      <button
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "marginRight": "5px",
          }
        }
        title="Přiřadí zvolenou masáž a otevře předdefinovaný editor Google událostí v nové záložce"
        type="button"
      >
        Zapsat a přidat do kalendáře
      </button>
    </a>
    <button
      autoFocus={false}
      className="btn btn-primary"
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "marginRight": "5px",
        }
      }
      title=""
      type="button"
    >
      Zapsat se
    </button>
    <button
      className="btn btn-default"
      onClick={[Function]}
      type="button"
    >
      Zrušit
    </button>
  </div>
</div>
`;
